Return extra metadata when a setting is controlled by a compliance policy

michalkleiner opened the pull request:
### Description:

When a setting is controlled by a compliance policy, it is handy to return that information in the API responses so that UI can optionally display such information to the admin section users.

The newly added short name getter methods have already deprecated and we will aim to make the existing protected name getters public. Not changing the visibility is solely required by an already released version of premium plugins. We can change that in the next major version.

Note: I've added the Matomo licence information to all the recently added files relating to the privacy policy compliance in a separate commit.

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

nathangavin commented on core/Settings/Interfaces/Traits/Getters/SystemGetterTrait.php:
```diff
@@ -35,7 +35,7 @@ public static function getSystemValue()
     abstract protected static function getSystemDefaultValue();
 
-    abstract protected static function getSystemName(): string;
+    abstract public static function getSystemName(): string;
```
We cant make this change without breaking the released version of HeatmapsSessionRecording plugin.

michalkleiner replied:
Reverted and used a workaround using a reflection. Also added a TODO and a deprecation to change this for Matomo 6.

michalkleiner commented:
@sgiehl keen to hear your opinion on the additions here, thanks!

With Nathan we do appreciate that using the reflection method mechanism isn't ideal, but with the existing release of Heatmaps plugin we think it's the best way forward for now. And then in Matomo 6 we can remove it and replace it with `getSettingName` public method on all the settings.

michalkleiner committed:
CS tweaks, update PHPStan baseline

michalkleiner committed:
Add custom names to IP anonymisation policy-managed settings

michalkleiner committed:
Add mechanism to get a name from policy-controlled setting

Using method reflection for plugins' backward compatibility.
The changes should be removed in Matomo 6 in favour of a public getSettingName method.

michalkleiner committed:
Add method to get a list of policies controlling a setting with its required value

michalkleiner committed:
Add setting/field metadata about compliance policy control and which site ID the setting relates to

michalkleiner committed:
Add system tests

sgiehl commented on core/Settings/Interfaces/Traits/PolicyComparisonTrait.php:
```diff
@@ -18,6 +18,7 @@ public static function getPolicyRequiredValues(?int $idSite = null): array
     {
         $policyValues = self::getPolicyRequirements();
 
+        /** @var class-string<CompliancePolicy> $policy */
```
This might be unneeded, considering that `getPolicyRequirements` already has a type hint, that defines that the array keys are of that type

michalkleiner replied:
Without that type hint, I get `Method 'isActive' not found in int|string` warning.

sgiehl replied:
I guess you mean within PHPStorm, right? PHPStan should be happy without it. But it's fine to keep it.

sgiehl commented on plugins/CorePluginsAdmin/tests/System/ApiTest.php:
```diff
@@ -0,0 +1,93 @@
+    }
+}
+
+ApiTest::$fixture = new PluginUpdatesFixture();
```
This could also use the `EmptySite` fixture I guess. There shouldn't be a need to fake a plugin update here, right?

michalkleiner replied:
good suggestion, changed, thanks!

sgiehl commented on core/Policy/PolicyManager.php:
```diff
@@ -109,4 +110,88 @@ public static function getAllUnknownSettings(string $policyClass): array
+        // if we found a method name, use reflection to make it accessible and then call it
+        if ($methodName) {
+            $reflection = new ReflectionMethod($controlledSettingClass, $methodName);
+            $reflection->setAccessible(true);
```
We shouldn't use this method. It has been deprecated with 8.5. See https://www.php.net/manual/en/reflectionproperty.setaccessible.php
Also I'm wondering why productive code might need to rely on such a method. If a method should be publicly accessible for any reason we should change the declaration. Or is anything preventing us from doing that?

nathangavin replied:
@sgiehl The reason we cant set these methods as public is because it would break the released version of the Heatmaps plugin.

@michalkleiner with regards to the above, I did some testing removing the usage of `setAccessible()` and it seems to not be necessary for the reflection to work.

michalkleiner replied:
Yeah, the whole mechanism using the `ReflectionMethod` was added to cater for an already released version of the Heatmaps plugin. You had that chat with Nathan on Slack. I've added a deprecation and todo to improve that for Matomo 6.

For now, I've added a version check to only call it for PHP < 8.1.

Happy to discuss other options if there are any.

sgiehl reviewed (changes requested):
Left some comments. Besides that I'm not fully sure if the additional data returned by the API will be sufficient for our needs in the UI as I didn't have time to look through all the other code around that topic. But for just showing a message it might be fine.

michalkleiner committed:
Only use setAccessible call before PHP 8.1

michalkleiner committed:
Use EmptySite fixture for system tests

michalkleiner commented:
> I'm not fully sure if the additional data returned by the API will be sufficient for our needs in the UI as I didn't have time to look through all the other code around that topic. But for just showing a message it might be fine.

Yep, I am aware of that we might need more information returned in the future if we want to implement some more smarts on the topic of conditionally disabling less strict options for some of the settings, but that was out of scope for the ticket. Potentially even the required value is redundant for now from that perspective, but it seems like a reasonable start.

michalkleiner committed:
Add public short name getters to getter traits

Using a map of getter traits and their methods to allow to obtain a subset of settings based
on their setting type (measurable, system, custom, etc.) and the context the method to compare
the setting name to a compliance policy-controlled setting name is called from.

michalkleiner committed:
Add licence information to recently added files

sgiehl commented on core/Policy/PolicyManager.php:
```diff
@@ -48,7 +95,7 @@ public static function getPolicyByName(string $policyName): ?string
      * @return array<class-string<PolicyComparisonInterface<mixed>&SettingValueInterface<mixed>>>
      */
-    public static function getAllSettings(?int $idSite = null): array
+    public static function getAllSettings(?int $idSite = null, ?string $settingType = null): array
```
Any reason why this has `$idSite` as param. It currently seems unused.

michalkleiner replied:
Probably doesn't need it as it's only called internally. Will remove.

sgiehl commented on core/Policy/PolicyManager.php:
```diff
@@ -109,4 +159,61 @@ public static function getAllUnknownSettings(string $policyClass): array
+            case MeasurableSetting::class:
+            case MeasurableProperty::class:
+                return self::SETTING_TYPE_MEASURABLE;
+            case SystemSetting::class:
```
For completeness we could also handle `SystemConfigSetting:class` and map it to config.

michalkleiner replied:
Good suggestion, will add, thanks

michalkleiner replied:
Refactored the method at the same time to support subclasses of settings, which previously prevented displaying providing the information to a premium plugin.

sgiehl reviewed (dismissed):
I hadn't done any of the reviews of the compliance stuff before, so it would be too much effort for me right not to do a proper in detail review, as I would first need to look at all the things that were done before.
However, I've left some comments that came to my mind when looking at the code.
From an API point of view the changes should be fine. Introducing the new name getter methods looks like a sensible approach now.
I didn't have time to test it with all plugins that already have settings classes, so I'm unable to say if there would be other problems, but I'm sure you checked that ;-)

michalkleiner committed:
Remove unused idsite param and change visibility to protected

michalkleiner committed:
Add SystemConfigSetting to setting type resolving

michalkleiner committed:
Refactor setting type getter method to support sublasses

nathangavin reviewed (approved):
Looks good to me, all CNIL changes are good and the general cleanup is appreciated.

nathangavin committed:
Merge branch '5.x-dev' into dev-19615

nathangavin commented on core/Policy/PolicyManager.php:
```suggestion
        $settings = static::getAllSettings();
```
hangover from other PR, needs to match new method param requirements.

michalkleiner committed:
Merge branch '5.x-dev' into dev-19615

michalkleiner committed:
Adjustments after 5.x-dev merged related CNIL work

michalkleiner committed:
Merge branch '5.x-dev' into dev-19615
